File: v2/AssembleTestable.java
public class AssembleTestable extends Assemble {
    private String[] inputs;
    private int inputIndex;
    private StringBuilder out;

    public String run(String[] inputs) {
        this.inputs = inputs;
        this.inputIndex = 0;
        this.out = new StringBuilder();

        execute();

        return out.toString();
    }

    public String run(String input) {
        if (input == null || input.isEmpty()) {
            return run(new String[0]);
        }

        return run(input.split("\\R"));
    }

    @Override
    protected void print(String msg) {
        out.append(msg);
    }

    @Override
    protected void println(String msg) {
        out.append(msg).append("\n");
    }

    @Override
    protected String readLine() {
        if (inputIndex >= inputs.length) {
            return null;
        }

        return inputs[inputIndex++];
    }

    @Override
    protected void clear() {
        // UnitTest에서는 화면 지우기 안 함
    }

    @Override
    protected void delay(int ms) {
        // UnitTest에서는 delay 안 함
    }
}

File: python/assemble_testable.py
from assemble import Assemble


class AssembleTestable(Assemble):
    def __init__(self):
        super().__init__()
        self._inputs: list[str] = []
        self._input_index = 0
        self._out: list[str] = []

    def run(self, inputs) -> str:
        if isinstance(inputs, str):
            inputs = inputs.splitlines() if inputs else []

        self._inputs = list(inputs)
        self._input_index = 0
        self._out = []

        self.execute()

        return "".join(self._out)

    def print_(self, msg: str) -> None:
        self._out.append(msg)

    def println(self, msg: str) -> None:
        self._out.append(msg + "\n")

    def read_line(self) -> str | None:
        if self._input_index >= len(self._inputs):
            return None

        line = self._inputs[self._input_index]
        self._input_index += 1
        return line

    def clear(self) -> None:
        pass

    def delay(self, ms: int) -> None:
        pass
